man/man3/mpl_div.man
.TH "mpl_div" "3" "23 ноября 2012" "Linux" "MPL Functions Manual"
.
.SH ИМЯ
mpl_add, mpl_sub, mpl_mul, mpl_div \- простейшие арифметические действия
.
.SH СИНТАКСИС
.nf
.B #include <mpl.h>
.sp
.BI "int mpl_add(mpl_int *" c ", const mpl_int *" a ", conts mpl_int *" b );
.br
.BI "int mpl_sub(mpl_int *" c ", const mpl_int *" a ", conts mpl_int *" b );
.br
.BI "int mpl_mul(mpl_int *" c ", const mpl_int *" a ", conts mpl_int *" b );
.br
.BI "int mpl_div(mpl_int *" q ", mpl_int *" r ", const mpl_int *" y ", conts mpl_int *" x );
.fi
.
.SH ОПИСАНИЕ
Функция \fBmpl_add\fP складывает \fIa\fP и \fIb\fP
и записывает сумму в \fIc\fP.
.br
.IR \tc " = " a " + " b
.P
Функция \fBmpl_sub\fP вычитает из \fIa\fP число \fIb\fP
и записывает разность в \fIc\fP.
.br
.IR \tc " = " a " - " b
.P
Функция \fBmpl_mul\fP умножает \fIa\fP на \fIb\fP
и записывает произведение в \fIc\fP.
.br
.IR \tc " = " a " * " b
.P
Функция \fBmpl_div\fP делит с остатком \fIy\fP на \fIx\fP
и записывает частное в \fIq\fP,
а остаток в \fIr\fP.
.br
.IR \tq " = " y " / " x
.br
.IR \tr " = " y " % " x
.P
Если остаток не нужен,
то на место \fIr\fP можно подставить NULL:
.nf
	mpl_int a, b, c;
.sp
	...
.sp
	mpl_div(&c, NULL, &a, &b);
.sp
	...
.fi
.
.SH "ВОЗВРАЩАЕМОЕ ЗНАЧЕНИЕ"
Все функции возвращают \fBMPL_OK\fP в случае успеха,
или \fBMPL_NOMEM\fP,
если произошла ошибка.
.P
Функция \fBmpl_div\fP возвращает \fBMPL_ERR\fP,
если делитель равен нулю.
.
.SH ЗАМЕЧАНИЯ
Все функции могут использовать переменные-операнды
как переменные для записи результата. К примеру:
.br
.nf
	mpl_int a, b;
.sp
	...
.sp
	mpl_mul(&b, &b, &b);
	mpl_add(&a, &a, &b);
	mpl_div(&b, &a, &a, &b);
.sp
	...
.fi
.
.SH "СМОТРИ ТАКЖЕ"
.BR mpl_set_one (3),
.BR mpl_set_sint (3),
.BR mpl_set_uint (3),
.BR mpl_set_uchar (3),
.BR mpl_set_str (3),
.BR mpl_cmp (3),
.BR mpl_abs_cmp (3),
.BR mpl_mul_dig (3),
.BR mpl_mul_ndig (3),
.BR mpl_sqr (3),
.BR mpl_gcd (3),
.BR mpl_primality_miller_rabin (3),
.BR mpl_reduce_barrett
.BR mpl_mod_exp (3),
.BR mpl_mod_inv (3),
.BR mpl_shr (3),
.BR mpl_shl (3),
.BR mpl_and (3),
.BR mpl_or (3),
.BR mpl_xor (3),
.BR mpl_copy (3),
.BR mpl_swap (3),
.BR mpl_to_uchar (3),
.BR mpl_to_str (3)


src/mpl_div.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

#include "mpl.h"
#include "mpl_common.h"

extern void mpl_canonicalize(mpl_int *a);

int
mpl_div(mpl_int *q, mpl_int *r, const mpl_int *y, const mpl_int *x)
{
	mpl_int u, v, tmp;
	_mpl_int_t dig;
	int norm, cnt, rc;
	int ytop, ysign;
	int xtop, xsign;
	int i, n, t, sign;

	if (mpl_iszero(x))
		return MPL_ERR;

	if (mpl_abs_cmp(x, y) == MPL_CMP_GT) {
		if (r != NULL) {
			if ((rc = mpl_copy(r, y)) != MPL_OK)
				return rc;
		}
		mpl_zero(q);
		return MPL_OK;
	}

	if ((rc = mpl_initv(&u, &v, &tmp, NULL)) != MPL_OK)
		return rc;

	if ((rc = mpl_copy(&u, y)) != MPL_OK)
		goto err;

	if ((rc = mpl_copy(&v, x)) != MPL_OK)
		goto err;

	u.sign = MPL_SIGN_POS;
	v.sign = MPL_SIGN_POS;

	ytop = y->top;
	xtop = x->top;
	ysign = y->sign;
	xsign = x->sign;

	/* Guess about the result sign. */
	sign = (ysign == xsign) ? MPL_SIGN_POS: MPL_SIGN_NEG;

	/* Reserve memory in quotient and remainder. */
	if ((rc = mpl_ensure(q, ytop - xtop + 1)) != MPL_OK)
		goto err;

	if (r != NULL) {
		if ((rc = mpl_ensure(r, xtop + 1)) != MPL_OK)
			goto err;
	}

	/* Count bits of the topmost digit of divider. */
	dig = x->dig[xtop];
	cnt = 0;
	while (dig > 0) {
		dig >>= 1;
		++cnt;
	}

	/* Normalize divider and dividend. */
	if (cnt < MPL_INT_BITS) {
		norm = (MPL_INT_BITS) - cnt;
		if ((rc = mpl_shl(&u, norm)) != MPL_OK)
			goto err;
		if ((rc = mpl_shl(&v, norm)) != MPL_OK)
			goto err;
	} else {
		norm = 0;
	}

	/* Setup quotient properties known in advance. */
	mpl_zero(q);
	q->top = ytop - xtop;
	q->sign = sign;

	for (;;) {
		n = u.top;
		t = v.top;
		if (n < t)
			break;
		cnt = (n - t) * MPL_INT_BITS;
		mpl_shl(&v, cnt);
		if (mpl_abs_cmp(&u, &v) != MPL_CMP_LT) {
			do {
				++q->dig[n-t];
				if ((rc = mpl_sub(&u, &u, &v)) != MPL_OK)
					goto err;
			} while (mpl_abs_cmp(&u, &v) != MPL_CMP_LT);
			mpl_shr(&v, cnt);
		} else {
			mpl_shr(&v, cnt);
			break;
		}
	}

	for (i = n; i >= t+1; i--) {
		_mpl_int_t tmpint[3];
		_mpl_int_t qcap;
		mpl_int tmp2;
		int qpos;

		/* Skip empty dividend digits. */
/*		if (i > u.top)
			continue;
*/
		/* Precalculate quotient digit position. */
		qpos = i - t - 1;

		/* First approximation to quotient. */
		if (u.dig[i] == v.dig[t]) {
			qcap = MPL_INT_BASE-1;
		} else {
			_mpl_long_t tmp;
			tmp  = (_mpl_long_t) u.dig[i] << MPL_INT_BITS;
			tmp |= (_mpl_long_t) u.dig[i-1];
			tmp /= (_mpl_long_t) v.dig[t];
			if (tmp >= MPL_INT_BASE)
				tmp = MPL_INT_BASE-1;
			qcap = tmp & MPL_INT_MASK;
		}

		++qcap;

		/* 
		 * Refine quotient approximation using 3 digits of dividend and
		 * 2 digits of divisor. This refinement ensures that quotient
		 * will be less then or equal to (q+1).
		 */
		do {
			--qcap;
			tmp.top = 1;
			tmp.sign = MPL_SIGN_POS;
			tmp.dig[1] = v.dig[t];
			tmp.dig[0] = t > 0 ? v.dig[t-1] : 0;

			tmp2.top = 2;
			tmp2.dig = tmpint;
			tmp2.sign = MPL_SIGN_POS;
			tmpint[2] = u.dig[i];
			tmpint[1] = i > 0 ? u.dig[i-1] : 0;
			tmpint[0] = i > 1 ? u.dig[i-2] : 0;

			if ((rc = mpl_mul_dig(&tmp, &tmp, qcap)) != MPL_OK)
				goto err;

			++cnt;
		} while (mpl_abs_cmp(&tmp, &tmp2) == MPL_CMP_GT);
	
		/* u = u - qv*b^{i-t-1} */

		if ((rc = mpl_mul_dig(&tmp, &v, qcap)) != MPL_OK)
			goto err;

		if ((rc = mpl_shl(&tmp, qpos*MPL_INT_BITS)) != MPL_OK)
			goto err;

		if ((rc = mpl_sub(&u, &u, &tmp)) != MPL_OK)
			goto err;

		if (u.sign == MPL_SIGN_NEG) {
			/* Fix divisor if the quotient approximation was by one bigger. */
			--qcap;
			if ((rc = mpl_mul_dig(&tmp, &v, qcap)) != MPL_OK)
				goto err;
			if ((rc == mpl_shl(&tmp, qpos*MPL_INT_BITS)) != MPL_OK)
				goto err;	
			if ((rc = mpl_add(&u, &u, &tmp)) != MPL_OK)
				goto err;
		}

		/* Store quotient digit. */
		q->dig[qpos] = qcap;
	}

	if (r != NULL) {
		mpl_copy(r, &u);
		if (norm > 0)
			mpl_shr(r, norm);
		r->sign = ysign;
	}

	mpl_canonicalize(q);
	rc = MPL_OK;
err:
	mpl_clearv(&u, &v, &tmp, NULL);
	return rc;
}

